File: App/src/main/java/it/gruppoinfor/home2work/fragments/MatchFragment.java
package it.gruppoinfor.home2work.fragments;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AnimationUtils;
import android.view.animation.LayoutAnimationController;
import android.widget.TextView;

import com.afollestad.materialdialogs.MaterialDialog;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import es.dmoral.toasty.Toasty;
import it.gruppoinfor.home2work.R;
import it.gruppoinfor.home2work.activities.MainActivity;
import it.gruppoinfor.home2work.activities.MatchActivity;
import it.gruppoinfor.home2work.activities.SettingsActivity;
import it.gruppoinfor.home2work.activities.ShowUserActivity;
import it.gruppoinfor.home2work.adapters.MatchAdapter;
import it.gruppoinfor.home2work.interfaces.ItemClickCallbacks;
import it.gruppoinfor.home2workapi.HomeToWorkClient;
import it.gruppoinfor.home2workapi.model.Match;
import it.gruppoinfor.home2workapi.model.User;

public class MatchFragment extends Fragment implements ItemClickCallbacks {

    @BindView(R.id.matches_recycler_view)
    RecyclerView matchesRecyclerView;
    @BindView(R.id.swipe_refresh_layout)
    SwipeRefreshLayout swipeRefreshLayout;
    @BindView(R.id.no_matches_view)
    TextView noMatchesView;
    private Unbinder mUnbinder;
    private MatchAdapter matchesAdapter;
    private Context mContext;

    private List<Match> matchList = new ArrayList<>();

    public MatchFragment() {
        // Required empty public constructor
    }

    @Override
    public void onAttach(Context context) {
        mContext = context;
        super.onAttach(context);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_match, container, false);
        mUnbinder = ButterKnife.bind(this, rootView);
        setHasOptionsMenu(true);
        swipeRefreshLayout.setColorSchemeResources(R.color.colorAccent);
        swipeRefreshLayout.setOnRefreshListener(() -> {
            swipeRefreshLayout.setRefreshing(true);
            refreshData();
        });
        return rootView;
    }

    @Override
    public void onResume() {
        super.onResume();
        refreshData();
    }

    @Override
    public void onItemClick(View view, int position) {
        Match match = matchList.get(position);
        if (match.getScore() == 0) {
            showMatchUserProfile(position);
        } else {
            if (match.isNew()) {
                match.setNew(false);
                matchesAdapter.notifyItemChanged(position);
                refreshBadgeCounter();
            }
            Intent matchIntent = new Intent(getContext(), MatchActivity.class);
            matchIntent.putExtra("match", matchList.get(position));
            startActivity(matchIntent);
        }
    }

    @Override
    public boolean onLongItemClick(View view, int position) {

        String[] options = new String[]{"Mostra profilo utente", "Nascondi"};

        new MaterialDialog.Builder(mContext)
                .items(options)
                .itemsCallback((dialog, itemView, p, text) -> {
                    switch (p) {
                        case 0:
                            showMatchUserProfile(position);
                            break;
                        case 1:
                            startActivity(new Intent(getActivity(), SettingsActivity.class));
                            break;
                        case 2:
                            showHideMatchDialog(position);
                            break;
                    }
                })
                .show();


        return true;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mUnbinder.unbind();
    }

    private void refreshData() {
        HomeToWorkClient.getInstance().getUserMatches(matches -> {
            matchList = matches;
            refreshBadgeCounter();
            refreshList();
            swipeRefreshLayout.setRefreshing(false);
        }, e -> {
            //Toasty.error(activity, "Impossibile ottenere i match").show();
            swipeRefreshLayout.setRefreshing(false);
        });
    }

    private void refreshList() {

        boolean noMatches = false;
        for (Match match : matchList) {
            if (match.getScore() == 0) noMatches = true;
        }

        if (noMatches) noMatchesView.setVisibility(View.VISIBLE);
        else noMatchesView.setVisibility(View.GONE);

        LinearLayoutManager layoutManager = new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false);
        LayoutAnimationController animation = AnimationUtils.loadLayoutAnimation(getContext(), R.anim.layout_animation);

        matchesRecyclerView.setLayoutManager(layoutManager);
        matchesRecyclerView.setLayoutAnimation(animation);

        matchesAdapter = new MatchAdapter(getActivity(), matchList);
        matchesAdapter.notifyDataSetChanged();
        matchesRecyclerView.setAdapter(matchesAdapter);
        matchesAdapter.setItemClickCallbacks(this);
    }

    protected void refreshBadgeCounter() {
        Integer newMatches = 0;

        for (Match match : matchList)
            if (match.isNew()) newMatches++;

        ((MainActivity) mContext).setBadge(1, newMatches > 0 ? newMatches.toString() : "");
    }

    private void showMatchUserProfile(int position) {
        Intent userIntent = new Intent(getActivity(), ShowUserActivity.class);
        User matchedUser = matchList.get(position).getHost();
        userIntent.putExtra("user", matchedUser);
        startActivity(userIntent);
    }

    private void showHideMatchDialog(int position) {
        Match matchItem = matchList.get(position);
        MaterialDialog hideDialog = new MaterialDialog.Builder(mContext)
                .title(R.string.item_match_dialog_hide_title)
                .content(R.string.item_match_dialog_hide_content)
                .positiveText(R.string.item_match_dialog_hide_confirm)
                .negativeText(R.string.item_match_dialog_hide_cancel)
                .onPositive((dialog, which) -> {

                    matchItem.setHidden(true);

                    HomeToWorkClient.getInstance().editMatch(matchItem, match -> {
                        matchList.remove(position);
                        matchesAdapter.remove(position);
                    }, e -> Toasty.success(mContext, mContext.getString(R.string.item_match_dialog_hide_error)).show());


                })
                .build();

        hideDialog.show();
    }


}

File: App/src/main/java/it/gruppoinfor/home2work/fragments/MatchFragment.kt
package it.gruppoinfor.home2work.fragments

import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.view.*
import android.view.animation.AnimationUtils
import android.widget.Toast
import com.afollestad.materialdialogs.MaterialDialog
import com.google.android.gms.tasks.OnFailureListener
import com.google.android.gms.tasks.OnSuccessListener
import it.gruppoinfor.home2work.R
import it.gruppoinfor.home2work.activities.*
import it.gruppoinfor.home2work.adapters.MatchAdapter
import it.gruppoinfor.home2work.interfaces.ItemClickCallbacks
import it.gruppoinfor.home2work.utils.Const
import it.gruppoinfor.home2workapi.HomeToWorkClient
import it.gruppoinfor.home2workapi.match.Match
import kotlinx.android.synthetic.main.fragment_match.*
import org.jetbrains.anko.intentFor
import org.jetbrains.anko.singleTop

class MatchFragment : Fragment() {

    private var matchesAdapter: MatchAdapter? = null
    private var matchList: ArrayList<Match> = ArrayList()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {

        val rootView = inflater.inflate(R.layout.fragment_match, container, false)
        setHasOptionsMenu(true)

        return rootView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        initUI()
        getMatches()

    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.menu_fragment_match, menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    private fun initUI() {

        swipe_refresh_layout.setColorSchemeResources(R.color.colorAccent)
        swipe_refresh_layout.setOnRefreshListener {
            swipe_refresh_layout.isRefreshing = true
            refreshMatches()
        }

        val layoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
        val animation = AnimationUtils.loadLayoutAnimation(context, R.anim.layout_animation)

        matches_recycler_view.layoutManager = layoutManager
        matches_recycler_view.layoutAnimation = animation

        matchesAdapter = MatchAdapter(context!!, matchList)

        matches_recycler_view.adapter = matchesAdapter

        matchesAdapter?.setItemClickCallbacks(object : ItemClickCallbacks {

            override fun onItemClick(view: View, position: Int) {

                val match = matchList[position]
                if (match.score == 0) {
                    showMatchUserProfile(position)
                } else {
                    if (match.isNew) {
                        match.isNew = false
                        matchesAdapter?.notifyItemChanged(position)
                        refreshBadgeCounter()
                    }
                    val matchIntent = Intent(context, MatchActivity::class.java)
                    matchIntent.putExtra(Const.EXTRA_MATCH, matchList[position])
                    startActivity(matchIntent)
                }

            }

            override fun onLongItemClick(view: View, position: Int): Boolean {

                val options = arrayOf("Mostra profilo utente", "Nascondi")

                MaterialDialog.Builder(context!!)
                        .items(*options)
                        .itemsCallback { _, _, p, _ ->
                            when (p) {
                                0 -> showMatchUserProfile(position)
                                1 -> startActivity(Intent(activity, SettingsActivity::class.java))
                                2 -> showHideMatchDialog(position)
                            }
                        }
                        .show()


                return true
            }

        })

    }

    private fun getMatches() {

        status_view.loading()

        HomeToWorkClient.getMatchList(OnSuccessListener { matches ->

            matchList.clear()
            matchList.addAll(matches)
            matchesAdapter?.notifyDataSetChanged()

            refreshUI()
            refreshBadgeCounter()

        }, OnFailureListener {

            status_view.error("Impossibile ottenere lista match al momento")
            it.printStackTrace()

        })

    }

    private fun refreshMatches() {

        HomeToWorkClient.getMatchList(OnSuccessListener { matches ->

            matchList.clear()
            matchList.addAll(matches)
            matchesAdapter?.notifyDataSetChanged()

            refreshBadgeCounter()
            refreshUI()

        }, OnFailureListener {

            swipe_refresh_layout.isRefreshing = false
            Toast.makeText(context, "Impossibile aggiornare lista match al momento", Toast.LENGTH_SHORT).show()
            it.printStackTrace()

        })

    }

    private fun refreshUI() {

        when {
            matchList.isEmpty() -> {
                status_view.empty("Non sono ancora disponibili dei match per te.")
            }
            matchList.any { it.score == 0 } -> {
                status_view.done()
                no_matches_view.visibility = View.VISIBLE
            }
            else -> {
                status_view.done()
                no_matches_view.visibility = View.GONE
            }
        }

        swipe_refresh_layout.isRefreshing = false

    }

    fun refreshBadgeCounter() {
        val newMatches = matchList.count { it.isNew }

        (context as MainActivity).setNavigationBadge(Const.MATCHES_TAB, if (newMatches > 0) newMatches.toString() else "")
    }

    private fun showMatchUserProfile(position: Int) {

        val matchedUser = matchList[position].host
        startActivity(context!!.intentFor<ShowUserActivity>(Const.EXTRA_USER to matchedUser).singleTop())

    }

    private fun showHideMatchDialog(position: Int) {

        val matchItem = matchList[position]
        val hideDialog = MaterialDialog.Builder(context!!)
                .title(R.string.item_match_dialog_hide_title)
                .content(R.string.item_match_dialog_hide_content)
                .positiveText(R.string.item_match_dialog_hide_confirm)
                .negativeText(R.string.item_match_dialog_hide_cancel)
                .onPositive { _, _ ->

                    matchItem.hidden = true

                    HomeToWorkClient.editMatch(matchItem, OnSuccessListener {
                        matchList.removeAt(position)
                        matchesAdapter?.remove(position)
                    }, OnFailureListener { Toast.makeText(context!!, R.string.item_match_dialog_hide_error, Toast.LENGTH_SHORT).show() })


                }
                .build()

        hideDialog.show()

    }


}

